### PR TITLE
Fix issue #60 : Wiki page explaining commonly encountered Server Errors and how to prevent them 

### PR DESCRIPTION
**Pull request by vaishnavi192**
Fix issue #60 
Added Server Errors and solutions page.
Updated sidebar with this page link to make it more discoverable to developers.

### CODE REVIEW

File: Server-errors-and-solutions.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+This document talks about different types of server problems that have been fixed before, and it suggests ways to stop those problems from happening again. The types of problems include issues with data, problems when lots of people are using the server, situations that don't happen very often, making sure the server works with old internet browsers, mistakes in the way the server thinks, problems when the internet is slow, and errors that don't need to be written down. For each type of problem, it gives an example and tips on how to stop it from happening.
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: document --> page

talks about --> discusses

I don't understand parts of the second sentence ("errors that don't need to be written down", "mistakes in the way the server thinks"). I suggest you just get rid of that sentence altogether.

File: Server-errors-and-solutions.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+* [Supporting old browsers](#supporting-old-browsers)
+* [Logic errors](#logic-errors)
+* [Data related issue](#data-related-issues)
+* [Race conditions / issues on slow networks](#race-conditions)
+* [Related to migration](#related-to-migrations)
+* [Errors that didn't need to be logged](#errors-that-didn't-need-to-be-logged)
+
+## Scalability Issue
+
+Issues that are related to features not scaling well on production. Example: (1) Exploration state data is too large(#12547) for some explorations and no further changes can be made to these explorations. (2) Suggestion models being fetched for review exceeds the query limit, so only an incomplete result is shown.
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: Link properly to the issue in question. Make sure the link works.

Put a space before the parentheses.

File: Server-errors-and-solutions.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+### Preventive action
+
+* __Design stage__: During design doc reviews, one should ask questions about performance and scalability when dealing with large amounts of data – “how big do you expect X to get?”, “are there any limits to the size?”, “what happens when it reaches this limit and is it possible to handle it?”
+* __Manual testing__: Having a functionality to generate dummy data for testing locally might help.
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: a --> some

File: Server-errors-and-solutions.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+## Corner Cases
+
+User-facing issues where certain corner case scenarios are not considered (empty list, undefined object etc.) Example: Deleting a hint that is active in the editor view caused console errors. Although these are not often serious, these issues can be tricky to debug due to lack of reproduction steps. The majority of server errors fall under this category. Other e.g. error #7907, #6298, #6199, #12528, #8702, #8464, #7970, #7546, #7444, #7727 (usage of plurals in certain languages), #7342, #7331, #7047, #6757, #6523 (didn't account for changes to interaction type), #5457
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: Make sure your links work (they currently don't). Use Markdown format for links.

Also, you don't need so many examples, at most 5 will do.

File: Server-errors-and-solutions.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+### Preventive action
+
+* __Static analysis / compilation stage__: Strongly typed frontend will help catch some of these issues.
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: Having strong typing in the frontend will...

File: Server-errors-and-solutions.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+* __Manual testing__: Developers generally test only the “happy” paths in their feature. We should encourage them to test edge cases. Adoption of template users in testing mentioned here should help uncover / test edge case scenarios.
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: The link for "here" is missing. Use https://docs.google.com/document/d/1ktMagEKvUA6A0UmM_2CoENZPelqmjLmRjV7tWeo6w54/edit#heading=h.y3evobat23c5

File: Server-errors-and-solutions.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+## Supporting old browsers
+Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other e.g. errors #6909, #6174, #7926, #7794, #7784, #7205
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: Other e.g. errors --> Other examples of errors include: ...


Also please fix the links, here and elsewhere (I won't comment on this any more).

File: Server-errors-and-solutions.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+## Logic errors
+Logical errors in the code. Example: “this” scope was referring to the global scope when it was supposed to refer to a local one (#7873). These are quite rare. Other e.g. errors #7873, #7872 (incorrect scope of this), #7927, #7460
+
+### Preventive action
+
+* __Code review__: These issues need to be flagged during code reviews. Often happens when the developer is refactoring existing code but hasn’t fully tested their changes.
+
+## Data-related Issue
+
+Issues that resulted in corrupt or missing data on the server. Issues like this are quite hard to debug and involve complicated one-off jobs to be run. Example: Certain exploration stats models were missing on the production server, this resulted in few deferred tasks erroring often and piling up in the stackdriver logs. Fortunately, these issues are quite rare. Other e.g. errors #6159, #9632, #7440
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: this resulted --> which resulted

few --> a few

stackdriver --> StackDriver

Please correct "Other e.g. errors" as mentioned above, and apply similar corrections to other sections.

File: Server-errors-and-solutions.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+### Preventive action
+
+* __Manual testing__: When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
+Testing with foreign / special characters (as mentioned under Template Users).
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: You should have 1., 2.  numbered bullet points here, similar to the original doc. Please preview your changes to make sure they are correct.

File: Server-errors-and-solutions.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+* __Review phase__: Reviewers should be encouraged to ask questions and ensure that the audit job is thorough.
+
+## Race Conditions / Slow Network Issues
+
+Some errors happen when interacting too fast with the application or when the network is too slow. Example: Clicking on the “content” tab in the translation editor before it has fully loaded (#6378). Other e.g. errors #6161, #13087, #8181, #8733, #8321, #7381, #7252, #7086, #7041, #6921, #6298, #6161   
+
+### Preventive action
+
+* __Manual testing__: For code that has a lot of “async” functionality, the authors should be encouraged to try throttling their network and/or interacting with the UI before it has completely loaded. This could be a category of “template users” mentioned here.
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: The link under "here" is missing. See above.

File: Server-errors-and-solutions.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+## Related to a migration
+
+One-off errors from migration of libraries or frameworks e.g. missing imports in the module files.
+
+### Preventive action
+* __Documentation__: Record common issues in the relevant migration page when they are identified.
+
+## Errors that didn't need to be logged
+Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors. Other e.g. errors: [$compile:tpload]" issue causing page load errors #5793, #6500, #8761, #7563
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: Why is there a $compile:tpload here?

File: Server-errors-and-solutions.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+### Preventive action
+* __Documentation__: Record common issues in the relevant migration page when they are identified.
+
+## Errors that didn't need to be logged
+Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors. Other e.g. errors: [$compile:tpload]" issue causing page load errors #5793, #6500, #8761, #7563 
+
+### Preventive action
+
+* __Code / review__: When adding logging statements or throwing errors that are not handled, always consider what scenarios and how often they will get called. If the frequency is high, discuss with the server error coordinator for the month to check if there are alternatives to achieve the same (e.g. it may be possible to add logs on stackdriver interface directly in a controlled way.)
```
Comment (seanlip) [inline Server-errors-and-solutions.md]: what scenarios --> what scenarios these arise in

same --> desired logging behaviour

**Review by seanlip — CHANGES_REQUESTED** [review]
Thanks @vaishnavi192, took a pass. Please make sure to proofread your work in preview and check that the links work before submitting a follow-up review request.

File: Server-errors-and-solutions.md (hunk 9 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: Done

File: Server-errors-and-solutions.md (hunk 1 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: I removed those 2 sentences for now. Do you want me to delete the whole sentence starting from "The type of problems"
till "written down"?

File: Server-errors-and-solutions.md (hunk 2 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: Yes sir done

File: Server-errors-and-solutions.md (hunk 3 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: Done

File: Server-errors-and-solutions.md (hunk 4 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: Now links are working

File: Server-errors-and-solutions.md (hunk 5 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 6 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: Ah, my bad I forgot. Corrected it now

File: Server-errors-and-solutions.md (hunk 7 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: Done

File: Server-errors-and-solutions.md (hunk 8 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: Done

File: Server-errors-and-solutions.md (hunk 10 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: Done

File: Server-errors-and-solutions.md (hunk 11 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: This was mentioned it in the word file provided in issue thread. I removed it now

File: Server-errors-and-solutions.md (hunk 12 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

**Comment by vaishnavi192** [comment]
@seanlip PTAL I did the required changes. Thanks

File: Server-errors-and-solutions.md (hunk 1 above)
Comment (seanlip) [reply Server-errors-and-solutions.md]: Yes, please do that.

**Review by seanlip — CHANGES_REQUESTED** [review]
@vaishnavi192 As far as I can tell, the links are not working. Please show a video where you can demonstrate this. I'm not going to review further until this is fixed.

**Comment by vaishnavi192** [comment]
> @vaishnavi192 As far as I can tell, the links are not working. Please show a video where you can demonstrate this. I'm not going to review further until this is fixed.

Yes I checked it, it is working properly. Have a look(Pointer is not visible in screencast tho)

[Screencast from 20-01-24 02:29:18 AM IST.webm](https://github.com/oppia/oppia-web-developer-docs/assets/117008958/bea26d22-8f0f-4c56-bef3-614b6b218e64)

File: Server-errors-and-solutions.md (hunk 1 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

**Comment by vaishnavi192** [comment]
@seanlip PTAL

**Comment by seanlip** [comment]
Due to review load I am going to defer follow-up review of this PR to @U8NWXD.

**Comment by vaishnavi192** [comment]
@U8NWXD PTAL

File: Server-errors-and-solutions.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+* [Supporting old browsers](#supporting-old-browsers)
+* [Logic errors](#logic-errors)
+* [Data related issue](#data-related-issue)
+* [Race Conditions / Slow Network Issues](#race-conditions--slow-network-issues)
+* [Related to migration](#related-to-migration)
+* [Errors that didn't need to be logged](#errors-that-didnt-need-to-be-logged)
+
+## Scalability Issue
+
+Issues that are related to features not scaling well on production. Example: (1) Exploration state data is too large (#12547) for some explorations and no further changes can be made to these explorations. (2) Suggestion models being fetched for review exceeds the query limit, so only an incomplete result is shown.
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]: As seanlip previously commented (https://github.com/oppia/oppia-web-developer-docs/pull/242#discussion_r1457818812), the issue link doesn't work. You can see this in the file preview here: https://github.com/oppia/oppia-web-developer-docs/blob/de2dfc2a44167ce568380c4d862021dfae0e6570/Server-errors-and-solutions.md. GitHub only auto-generates links for issues/PRs in issue, PR, and Discussion threads, not in general Markdown files or in the wiki.

In your video in https://github.com/oppia/oppia-web-developer-docs/pull/242#issuecomment-1901101129, you can see that the issue references aren't made into links.

Ditto elsewhere

File: Server-errors-and-solutions.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+### Preventive action
+
+1. __Static analysis / compilation stage__: Having strong typing in the frontend will help catch some of these issues.
+2. __Manual testing__: Developers generally test only the “happy” paths in their feature. We should encourage them to test edge cases. Adoption of template users in testing mentioned [here](https://docs.google.com/document/d/1ktMagEKvUA6A0UmM_2CoENZPelqmjLmRjV7tWeo6w54/edit#heading=h.y3evobat23c5) should help uncover / test edge case scenarios.
+
+## Supporting old browsers
+
+Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other examples of errors include: #6909, #6174, #7926, #7794
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]:
```suggestion
Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other examples of errors include: #6909, #6174, #7926, #7794.
```

Ditto elsewhere for ending sentences with punctuation.

File: Server-errors-and-solutions.md (hunk 9 above)
Comment (U8NWXD) [reply Server-errors-and-solutions.md]: not done

File: Sidebar.md [hunk 15]
```diff
@@ -42,6 +42,7 @@
     * [[How to write frontend type definitions|Guide-on-defining-types]]
     * [[How to write design docs|Writing-design-docs]]
     * [[Revert and Regression Policy|Revert-and-Regression-Policy]]
+    * [[Coding safely to avoid server errors|Coding-safely-to-avoid-server-errors]]
```
Comment (U8NWXD) [inline Sidebar.md]:
```suggestion
    * [[Server errors and solutions|Server-errors-and-solutions]]
```

This needs to match the file name, otherwise the link won't work

File: Server-errors-and-solutions.md [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+* [Scalability issue](#scalability-issue)
+* [Corner cases](#corner-cases)
+* [Supporting old browsers](#supporting-old-browsers)
+* [Logic errors](#logic-errors)
+* [Data related issue](#data-related-issue)
+* [Race Conditions / Slow Network Issues](#race-conditions--slow-network-issues)
+* [Related to migration](#related-to-migration)
+* [Errors that didn't need to be logged](#errors-that-didnt-need-to-be-logged)
+
+## Scalability Issue
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]: Headings should be sentence-case (i.e. "Issue" -> "issue"). Ditto elsewhere

File: Server-errors-and-solutions.md [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+Issues that are related to features not scaling well on production. Example: (1) Exploration state data is too large (#12547) for some explorations and no further changes can be made to these explorations. (2) Suggestion models being fetched for review exceeds the query limit, so only an incomplete result is shown.
+
+### Preventive action
+
+1. __Design stage__: During design doc reviews, one should ask questions about performance and scalability when dealing with large amounts of data – “how big do you expect X to get?”, “are there any limits to the size?”, “what happens when it reaches this limit and is it possible to handle it?”
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]: Likely because you copied text from the doc, you have curly quotes (`“`) which curl the wrong way. Please replace them with straight quotes (`"`), which is what usually gets generated when you type quotation marks. Programs like Google Docs and Word just automatically replace straight quotes with the curly ones.

Ditto elsewhere

File: Server-errors-and-solutions.md [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+2. __Manual testing__: Having some functionality to generate dummy data for testing locally might help.
+
+## Corner Cases
+
+User-facing issues where certain corner case scenarios are not considered (empty list, undefined object etc.) Example: Deleting a hint that is active in the editor view caused console errors. Although these are not often serious, these issues can be tricky to debug due to lack of reproduction steps. The majority of server errors fall under this category. Other examples of errors include: #7907, #7727 (usage of plurals in certain languages), #7342, #6523 (didn't account for changes to interaction type), #5457
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]:
```suggestion
User-facing issues where certain corner case scenarios are not considered (e.g. empty list or undefined object). Example: Deleting a hint that is active in the editor view caused console errors. Although these are not often serious, these issues can be tricky to debug due to lack of reproduction steps. The majority of server errors fall under this category. Other examples of errors include: #7907, #7727 (usage of plurals in certain languages), #7342, #6523 (didn't account for changes to interaction type), #5457
```

Re-structure sentence to avoid ending with "etc." since ending with a period-terminated abbreviation creates ambiguity about whether the sentence has ended

File: Server-errors-and-solutions.md [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+## Supporting old browsers
+
+Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other examples of errors include: #6909, #6174, #7926, #7794
+
+### Preventive action
+
+1. __Static analysis / compilation stage__: Fix a JS standard that we expect all clients to support and specify a “target version” in typescript so that the code will be compiled to that version / JS standard.
+
+## Logic errors
+Logical errors in the code. Example: “this” scope was referring to the global scope when it was supposed to refer to a local one (#7873). These are quite rare. Other examples of errors include: #7873, #7872 (incorrect scope of this), #7927, #7460
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]:
```suggestion
Logical errors in the code. Example: The “this” scope was referring to the global scope when it was supposed to refer to a local one (#7873). These are quite rare. Other examples of errors include: #7873, #7872 (incorrect scope of "this"), #7927, #7460
```

File: Server-errors-and-solutions.md [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+### Preventive action
+
+1. __Code review__: These issues need to be flagged during code reviews. Often happens when the developer is refactoring existing code but hasn’t fully tested their changes.
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]:
```suggestion
1. __Code review__: These issues need to be flagged during code reviews. They often happen when a developer is refactoring existing code but hasn’t fully tested their changes.
```

File: Server-errors-and-solutions.md [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+
+## Data-related Issue
+
+Issues that resulted in corrupt or missing data on the server. Issues like this are quite hard to debug and involve complicated one-off jobs to be run. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: #6159, #9632, #7440
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]:
```suggestion
Issues that result in corrupt or missing data on the server. Issues like this are quite hard to debug and involve running complicated one-off jobs. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: #6159, #9632, #7440
```

Let's keep the issue descriptions in present tense. Please also proofread for unnecessarily convoluted language like "complicated one-off jobs to be run"

File: Server-errors-and-solutions.md [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+
+### Preventive action
+
+1. __Manual testing__: When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
+Testing with foreign / special characters (as mentioned under Template Users).
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]: Not clear what "Template Users" is referring to. Do you need to link to a doc?

File: Server-errors-and-solutions.md [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+## Related to migration
+
+One-off errors from migration of libraries or frameworks e.g. missing imports in the module files.
+
+### Preventive action
+
+1. __Documentation__: Record common issues in the relevant migration page when they are identified.
+
+## Errors that didn't need to be logged
+Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors. Other examples of errors include: #5793, #6500, #8761, #7563
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]:
```suggestion

Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors. Other examples of errors include: #5793, #6500, #8761, #7563 
```

File: Server-errors-and-solutions.md [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+### Preventive action
+
+1. __Documentation__: Record common issues in the relevant migration page when they are identified.
+
+## Errors that didn't need to be logged
+Sometimes there may be unexpected errors thrown that do not impact user actions. Such errors can be ignored on a case-by-case basis. Example: Navigating away from a page when there are pending requests or if the request timed out caused some console errors. Other examples of errors include: #5793, #6500, #8761, #7563 
+
+### Preventive action
+
+1. __Code / review__: When adding logging statements or throwing errors that are not handled, always consider what scenarios these arise in and how often they will get called. If the frequency is high, discuss with the server error coordinator for the month to check if there are alternatives to achieve the desired logging behaviour (e.g. it may be possible to add logs on stackdriver interface directly in a controlled way.)
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]: stackdriver -> StackDriver

File: Server-errors-and-solutions.md [hunk 25]
```diff
@@ -0,0 +1,83 @@
+This page discusses different types of server problems that have been fixed before, and it suggests ways to stop those problems from happening again. For each type of problem, it gives an example and tips on how to stop it from happening.
+
+# Categories
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]:
```suggestion
## Table of contents
```

**Review by U8NWXD — CHANGES_REQUESTED** [review]

File: Server-errors-and-solutions.md (hunk 25 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 16 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 13 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: oh that issue. Now I get it I thought "scalability issues" link was not working. sorry
I linked issues with links now. I removed some issues whose links I was not able to find.

File: Server-errors-and-solutions.md (hunk 17 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 18 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 14 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 19 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 20 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 21 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 9 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: It didn't have any heading so I thought it will be another new line in manual testing para itself. 
okay Now I changed it

File: Server-errors-and-solutions.md (hunk 22 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: yeah done

File: Server-errors-and-solutions.md (hunk 23 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 24 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: I perfectly remember making these changes before when seanlip told. Don't know why it didn't get updated.
I did it again.

File: Sidebar.md (hunk 15 above)
Comment (vaishnavi192) [reply Sidebar.md]: done

**Comment by vaishnavi192** [comment]
@U8NWXD PTAL

File: Server-errors-and-solutions.md [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+
+## Data-related issue
+
+Issues that result in corrupt or missing data on the server. Issues that result in corrupt or missing data on the server. Issues like this are quite hard to debug and involve running complicated one-off jobs. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: [#6159](https://github.com/oppia/oppia/pull/6159), [#9632](https://github.com/oppia/oppia/pull/9632), [#7440](https://github.com/oppia/oppia/pull/7440).
+
+### Preventive action
+
+1. __Manual testing__: When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
+2. Testing with foreign / special characters (as mentioned under [Template Users](https://docs.google.com/document/d/1FYShgUHI0GM2I6eWe5B78PAQLvPxZxMFVjTivVoBrTg/edit#)).
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]: This doesn't match the [original doc](https://docs.google.com/document/d/1hIQIaFNVvyNiZBcCx0X7-_2lc0r1yvjxGfn4GJMjwEs/edit#heading=h.jlsplhp0gps7), and it's a little confusing. Performing destructive operations and using foreign/special characters are two types of manual testing that could help prevent this issue, so I think we actually want something like this:

> 1. __Manual testing__:
>    1. When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
>    2. Testing with foreign / special characters (as mentioned under [Template Users](https://docs.google.com/document/d/1FYShgUHI0GM2I6eWe5B78PAQLvPxZxMFVjTivVoBrTg/edit#)).

File: Server-errors-and-solutions.md [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+## Supporting old browsers
+
+Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other examples of errors include: [#6909](https://github.com/oppia/oppia/pull/6909), [#6174](https://github.com/oppia/oppia/pull/6174), [#7926](https://github.com/oppia/oppia/pull/7926).
+
+### Preventive action
+
+1. __Static analysis / compilation stage__: Fix a JS standard that we expect all clients to support and specify a "target version" in typescript so that the code will be compiled to that version / JS standard.
+
+## Logic errors
+Logical errors in the code. Example: The "this" scope was referring to the global scope when it was supposed to refer to a local one ([#7873](https://github.com/oppia/oppia/pull/7873)). These are quite rare. Other examples of errors include: [#7872](https://github.com/oppia/oppia/issues/7872)(incorrect scope of "this"), [#7927](https://github.com/oppia/oppia/pull/7927).
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]:
```suggestion
Logical errors in the code. Example: The "this" scope was referring to the global scope when it was supposed to refer to a local one ([#7873](https://github.com/oppia/oppia/pull/7873)). These are quite rare. Other examples of errors include: [#7872](https://github.com/oppia/oppia/issues/7872) (incorrect scope of "this"), [#7927](https://github.com/oppia/oppia/pull/7927).
```

Please proofread the rendered version to catch typos like this

File: Server-errors-and-solutions.md [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+
+### Preventive action
+
+1. __Code review__: These issues need to be flagged during code reviews. They often happen when a developer is refactoring existing code but hasn’t fully tested their changes.
+
+## Data-related issue
+
+Issues that result in corrupt or missing data on the server. Issues that result in corrupt or missing data on the server. Issues like this are quite hard to debug and involve running complicated one-off jobs. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: [#6159](https://github.com/oppia/oppia/pull/6159), [#9632](https://github.com/oppia/oppia/pull/9632), [#7440](https://github.com/oppia/oppia/pull/7440).
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]:
```suggestion
Issues that result in corrupt or missing data on the server. Issues like this are quite hard to debug and involve running complicated one-off jobs. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: [#6159](https://github.com/oppia/oppia/pull/6159), [#9632](https://github.com/oppia/oppia/pull/9632), [#7440](https://github.com/oppia/oppia/pull/7440).
```

**Review by U8NWXD — CHANGES_REQUESTED** [review]
Almost there! Just a few remaining comments. Also, remember to fix the merge conflict, otherwise we can't merge the PR

File: Server-errors-and-solutions.md (hunk 27 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: but their shouldn't be a gap between issue no. and bracket right?  I mean in general we write it as for eg.
28(twenty eight). So should I change it?

File: Server-errors-and-solutions.md (hunk 28 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: oh no, done

File: Server-errors-and-solutions.md (hunk 26 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

**Comment by vaishnavi192** [comment]
@U8NWXD PTAL I did the changes. Some changes were already updated locally and I pushed it before but they weren't there in previous branch changes. What could be the issue? 
Just wanted to ask do I have to unassign myself everytime I make changes?

**Comment by seanlip** [comment]
@vaishnavi192 Yes, please unassign yourself if you have completed your side of things for the PR.

I'm not sure why your changes were not reflecting, but that's another reason to always check the "Files Changed" tab before you resubmit a review request.

**Comment by vaishnavi192** [comment]
> @vaishnavi192 Yes, please unassign yourself if you have completed your side of things for the PR.
> 
> I'm not sure why your changes were not reflecting, but that's another reason to always check the "Files Changed" tab before you resubmit a review request.

Got it sir...Thanks! and sorry for creating mess😅... Will try to work more efficiently from next time.

File: Server-errors-and-solutions.md [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+## Data-related issue
+
+Issues that result in corrupt or missing data on the server. Issues like this are quite hard to debug and involve running complicated one-off jobs. Example: Certain exploration stats models were missing on the production server, which resulted in a few deferred tasks erroring often and piling up in the StackDriver logs. Fortunately, these issues are quite rare. Other examples of errors include: [#6159](https://github.com/oppia/oppia/pull/6159), [#9632](https://github.com/oppia/oppia/pull/9632), [#7440](https://github.com/oppia/oppia/pull/7440).
+
+### Preventive action
+
+1. __Manual testing__:  
+   i\. When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
+
+   ii\. Testing with foreign / special characters (as mentioned under [Template Users](https://docs.google.com/document/d/1FYShgUHI0GM2I6eWe5B78PAQLvPxZxMFVjTivVoBrTg/edit#)).
```
Comment (U8NWXD) [inline Server-errors-and-solutions.md]: You don't need to write out the `i` and `ii`. If you put in `1.` and `2.` just like any other numbered list, the rendered page will automatically transform to `i` and `ii`. As an additional benefit, the wrapped lines will be indented properly

File: Server-errors-and-solutions.md (hunk 27 above)
Comment (U8NWXD) [reply Server-errors-and-solutions.md]: There should be a gap between the issue number and parenthesis, and your example should be written as "28 (twenty eight)" with a space.

The only time I can think of where there wouldn't be a space before a parenthesis would be if you had a hierarchy of sections in a document. For example, if a question on an exam had a question number 1 with a subpart a, you could refer to the subpart as 1(a).

**Comment by U8NWXD** [comment]
> Some changes were already updated locally and I pushed it before but they weren't there in previous branch changes. What could be the issue?

This could be because after making the changes locally, you forgot to commit them. When you do a `git push`, only committed changes are uploaded

File: Server-errors-and-solutions.md (hunk 27 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: done

File: Server-errors-and-solutions.md (hunk 29 above)
Comment (vaishnavi192) [reply Server-errors-and-solutions.md]: okay done

**Comment by vaishnavi192** [comment]
@U8NWXD PTAL

**Review by U8NWXD — APPROVED** [review]

**Comment by U8NWXD** [comment]
Merging despite @seanlip's requested changes because I took over the review from seanlip
